### DISCUSSION TITLE
Strange transitions on pagination

### DISCUSSION THREAD

**Discussion by KattAttack** [pmndrs/react-spring#1325] [Support]
After loading enough kitties for several pages, the images appear to be sliding from right to left when switching between pages. I'm not sure what's causing this and would appreciate another set of eyes from someone who's more knowledgeable of this library. My goal is to have the images remain stationary when switching from one page to another (it's alright if opacity is affected).

My hypothesis is that the images from the previous page are disappearing slower than the new images on the page you've clicked on appear. So the new image appears before the others have fully disappeared, hence the delayed shift to the front of the box. But if this is true, is there a way to remedy this on my end?

https://codesandbox.io/s/zw1190rl2m

Thanks c:
-Kat

**Comment by harshagarwalnyu** [comment]
The sliding effect is caused by enter/leave animations overlapping — old items are still animating out while new items animate in, creating a visual "shift" as they compete for space.

**Fix: force exit before enter with `exitBeforeEnter` (react-spring v9+) or adjust timing:**

```jsx
import { useTransition, animated } from '@react-spring/web';

function PaginatedGallery({ images, page }) {
  const transitions = useTransition(images, {
    keys: (img) => `${page}-${img.id}`,  // key includes page — forces full remount on page change
    from: { opacity: 0 },
    enter: { opacity: 1 },
    leave: { opacity: 0 },
    // Option 1: exit before enter (sequential)
    exitBeforeEnter: true,
    // Option 2: just use opacity, no position animation
  });

  return transitions((style, img) => (
    <animated.img src={img.src} style={{ ...style, position: 'absolute' }} />
  ));
}
```

**The key insight:** including `page` in the key (`${page}-${img.id}`) ensures react-spring treats page-2's `img-1` as a completely different element from page-1's `img-1`. Without it, react-spring tries to animate the same key from old position → new position, causing the slide.

**If you want images to be stationary (only opacity):**
```js
from: { opacity: 0 },
enter: { opacity: 1 },
leave: { opacity: 0 },
// NO x/y/transform in the config
```

Remove any position-related animation values. The slide you're seeing is likely a leftover `x` or `transform: translateX` in your config.

**For `trail` staggering on enter:**
```js
trail: 50,  // 50ms delay between each item entering
```
